core/blockchain/blockchain_gateway.py

```diff
@@ -0,0 +1,355 @@
+    format='[BlockchainGateway] %(message)s')
+
+
+class BlockchainGateway(object):
+    """
+    Blockchain Gateway 
+
+    The blockchain client exposes `setter` and `getter` in order to interact
+    with the blockchain.
+
+    In order for this to work, the following must be running:
+        IPFS Daemon: `ipfs daemon`
+        The lotion app: `node app_trivial.js` from the application root directory
+
+    """
+
+    def __init__(self, config_manager, communication_manager):
+        """
+        TODO: Refactor dependencies
+        TODO: deal with config
+        """
+        # TODO: `communication_manager` is only used in a subset of methods,
+        # consider separating
+        self.communication_manager = communication_manager
+        # TODO: We will do this once the Communication Manager is done.
+        # self.communication_manager.configure_listener(self)
+
+        config = config_manager.get_config() if config_manager else {}
+        self.state = []
+        self.host = config.get("BLOCKCHAIN", "host")
+        self.ipfs_port = config.get("BLOCKCHAIN", "ipfs_port")
+        self.port = config.get("BLOCKCHAIN", "http_port")
+        self.client = None
+        try:
+            self.client = ipfsapi.connect(self.host, self.ipfs_port)
+        except Exception as e:
+            logging.info("IPFS daemon not started, got: {0}".format(e))
+
+        self.CALLBACKS = {
+            MessageEventTypes.NEW_WEIGHTS.name: self.broadcast_new_weights, 
+            MessageEventTypes.NOTHING.name: self._do_nothing,
+        }
+
+    ##########################################################################
+    ###                            API SECTION                             ###
+    ##########################################################################
+
+    def _do_nothing(self, payload: dict):
+        """
+        Do nothing.
+        """
+        pass
+
+    def get_global_state(self):
+        """
+        Gets the global state which should be a list of dictionaries
+        """
+        timeout = time.time() + 5
+        tx_receipt = None
+        while time.time() < timeout:
+            try:
+                tx_receipt = requests.get(
+                    "http://localhost:{0}/state".format(self.port))
+                tx_receipt.raise_for_status()
+                if tx_receipt:
+                    break
+            except (UnboundLocalError, requests.exceptions.ConnectionError) as e:
+                logging.info("HTTP GET error, got: {0}".format(e))
+                continue
+        logging.info("global state:{}".format(tx_receipt.json()))
+        return tx_receipt.json()
+
+    def get_diffs(self, old_state: list, new_state: dict) -> list:
+        """
+        Iterate through oldState and newState to see any differences
+        Take action based on the differences
+        """
+        tx_diffs = [tx for tx in new_state.get('messages') if (
+                        tx not in old_state)]
```

Comment:
Seems like you want to get all the elements in `new_state.get('messages')` that are not in `old_state`. Nitpick, but I would suggest using's `numpy`'s `setdiff1d`, i.e.
```
tx_diffs = np.setdiff1d(new_state.get('messages'), old_state)
```
`numpy` is vectorized, so if the size of the lists are large enough, the speedup may be signficant.

Reply:
lists are gonna be like 1mn+ long